### ISSUE TITLE
Build error running make on CentOS 7 (error: constructor required before non-static data member)

### ISSUE THREAD

**Issue by swoolbean** [cmu-sei/pharos#257]
Hi!
Hoping this is the right place to ask about this. I followed the installation guide for manually building from source to get Pharos working on my CentOS 7 virtual machine. As far as I can tell, I've managed to get all the dependencies built and installed, but when I run `make -j4` in Pharos' build directory, I get this error about halfway through the build:

```
/opt/rh/devtoolset-7/root/usr/include/c++/7/type_traits:878:48: error: constructor required before non-static data member for ‘pharos::PathFinder::SolverDeleter::owned’ has been parsed
     template<typename _Tp, typename = decltype(_Tp())>
```

with `make` finishing with this output:
```
make[2]: *** [libpharos/CMakeFiles/pharos.dir/options.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [libpharos/CMakeFiles/pharos.dir/path.cpp.o] Error 1
make[1]: *** [libpharos/CMakeFiles/pharos.dir/all] Error 2
make: *** [all] Error 2
```

Because I previously had an issue with CMake targeting an older version of C++, I also tried running it with extra flags to lock the version down, but I got the same error:

```sudo cmake -DCMAKE_INSTALL_PREFIX=/usr/local .. -DCMAKE_CXX_STANDARD=14 -DCMAKE_CXX_STANDARD_REQUIRED=ON```

I know that CentOS 7 isn't in your list of supported distributions for manually building, but RHEL 7 is, and in my experience that's a good indicator that CentOS 7 should be fine as well.

Please let me know if you need any more info or if I should be posting this somewhere else. Thanks!

**Comment by sei-eschwartz** [comment]
What version of g++ are you using? This is the right place to ask!

Sent from my T-Mobile 5G Device
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: SwoolBean ***@***.***>
Sent: Friday, September 15, 2023 3:24:11 AM
To: cmu-sei/pharos ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [cmu-sei/pharos] Build error running make on CentOS 7 (error: constructor required before non-static data member) (Issue #257)


Hi!
Hoping this is the right place to ask about this. I followed the installation guide for manually building from source to get Pharos working on my CentOS 7 virtual machine. As far as I can tell, I've managed to get all the dependencies built and installed, but when I run make -j4 in Pharos' build directory, I get this error about halfway through the build:

/opt/rh/devtoolset-7/root/usr/include/c++/7/type_traits:878:48: error: constructor required before non-static data member for ‘pharos::PathFinder::SolverDeleter::owned’ has been parsed
     template<typename _Tp, typename = decltype(_Tp())>


with make finishing with this output:

make[2]: *** [libpharos/CMakeFiles/pharos.dir/options.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [libpharos/CMakeFiles/pharos.dir/path.cpp.o] Error 1
make[1]: *** [libpharos/CMakeFiles/pharos.dir/all] Error 2
make: *** [all] Error 2


Because I previously had an issue with CMake targeting an older version of C++, I also tried running it with extra flags to lock the version down, but I got the same error:

sudo cmake -DCMAKE_INSTALL_PREFIX=/usr/local .. -DCMAKE_CXX_STANDARD=14 -DCMAKE_CXX_STANDARD_REQUIRED=ON

I know that CentOS 7 isn't in your list of supported distributions for manually building, but RHEL 7 is, and in my experience that's a good indicator that CentOS 7 should be fine as well.

Please let me know if you need any more info or if I should be posting this somewhere else. Thanks!

**Comment by swoolbean** [comment]
I want to say 7.2.4, but my host machine actually bricked itself this morning, so I won't be able to check for a few weeks until it's repaired 😅

If I have some time this coming week, I'll try getting it set up on a different PC.

**Comment by sei-mwd** [comment]
C++14 is supposedly available from gcc 6.1 and up, but we haven't tested with versions of gcc that old in some time.  It is possible that we accidentally have some code that is not C++14 compliant, but that compiles on later versions of gcc anyway.  I could look at what is going on, but I would need more context from your original error message to trace down what is happening.

**Comment by sei-mwd** [comment]
Did not mean to close this issue.

**Comment by sei-eschwartz** [comment]
I was able to build successfully using gcc 7.3.1 on centos 7:

https://colab.research.google.com/drive/1RsbkTSUjuIlrdR4T73vo4mC-K0Adifq0?usp=sharing

**Comment by sei-eschwartz** [comment]
https://stackoverflow.com/a/17436088

**Comment by sei-eschwartz** [comment]
This fixed it for me, but @sei-mwd might have a more elegant solution:

```
[root@ed-Precision-7540 libpharos]# git diff
diff --git a/libpharos/path.hpp b/libpharos/path.hpp
index a89b20a..2b31766 100644
--- a/libpharos/path.hpp
+++ b/libpharos/path.hpp
@@ -297,6 +297,8 @@ class PathFinder : public Z3PathAnalyzer {
   struct SolverDeleter {
     bool owned = true;
     void operator()(PharosZ3Solver *s) const { if (owned) delete s; }
+    SolverDeleter() {}
+    SolverDeleter(bool owned_) : owned(owned_) {}
   };
 
   // We will need Z3 for this analysis
```

The default constructor is needed for unique_ptr, and the second one is needed because we use `SolverDeleter{false}` somewhere.

**Comment by turol** [comment]
Single-parameter constructor should probably be marked `explicit`.

**Comment by sei-mwd** [comment]
Okay.  This looks like a bug in the C++ standard library in that version of gcc.  The fix that @sei-eschwartz posted is harmless, so I'll commit a variant of it shortly.

**Comment by sei-eschwartz** [comment]
@sei-mwd Can you make sure this is committed?

**Comment by sei-mwd** [comment]
I pushed before testing, and there was a typo.  Should be fixed in eeb1b52aac12e357db61f38ee0de022f1df881f8.
